apps/chat/src/components/Chat/ChatMessage/ChatMessageTemplatesModal/ChatMessageTemplatesModal.tsx

```diff
@@ -0,0 +1,244 @@
+import { IconChevronDown } from '@tabler/icons-react';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+
+import { useTranslation } from 'next-i18next';
+
+import classNames from 'classnames';
+
+import { templateMatchContent } from '@/src/utils/app/prompts';
+
+import { Conversation, Message } from '@/src/types/chat';
+import { ModalState } from '@/src/types/modal';
+import { Translation } from '@/src/types/translation';
+
+import { ConversationsActions } from '@/src/store/conversations/conversations.reducers';
+import { useAppDispatch } from '@/src/store/hooks';
+
+import { PROMPT_VARIABLE_REGEX } from '@/src/constants/folders';
+
+import Modal from '@/src/components/Common/Modal';
+
+import { TabButton } from '../../../Buttons/TabButton';
+import { TemplateRenderer } from './TemplateRenderer';
+import { TemplateRow } from './TemplateRow';
+
+interface Props {
+  isOpen: boolean;
+  onClose: (result: boolean) => void;
+  message: Message;
+  conversation: Conversation;
+}
+
+const EMPTY_ROW = ['', ''];
+
+export const ChatMessageTemplatesModal = ({
+  isOpen,
+  onClose,
+  message,
+  conversation,
+}: Props) => {
+  const { t } = useTranslation(Translation.Chat);
+  const dispatch = useAppDispatch();
+  const showMore = message.content.length > 160;
+  const [collapsed, setCollapsed] = useState(showMore);
+  const [previewMode, setPreviewMode] = useState(false);
+  const [templates, setTemplates] = useState([
+    ...Object.entries(message.templateMapping ?? {}),
+    EMPTY_ROW,
+  ]);
+
+  useEffect(() => {
+    if (isOpen) {
+      setTemplates([
+        ...Object.entries(message.templateMapping ?? {}),
+        EMPTY_ROW,
+      ]);
+    }
+  }, [message.templateMapping, isOpen]);
+
+  const handleChangeTemplate = useCallback(
+    (index: number, content: string, template: string) => {
+      const newTemplates = [...templates];
```

Comment:
Here is shallow copy of `templates`, I think we need to use `deepCopy` here to create new array

Reply:
We don't need `deepCopy` here